### PR TITLE
feat(Typography)!: Heading Primitives and reconfigured responsive styles

### PR DESCRIPTION
**Pull request by codecaaron**
# Before reviewing
The current props represent the baseline needs for both `Heading` and `Text` to be used effectively, but not necessarily the ideal state. All ideas are welcome on which tags / configurations we use and what level of inheritance we would like to support. 

An example usage in the wild: https://github.com/codecademy-engineering/Codecademy/pull/17017

## Notes
- Documentation still a work in progress.  Want to sync with design on best standards for this.

## Added
- New `Heading` primitive that allows for any h1-h6 and any configuration from xxs-xl.
- New `Text` primitive for (`p`'s and `span`s for now).

## Changes
- Moves mixins and screen size variables to gamut styles.
- Refactors classname mapping utility to be cleaner and produce less garbage classnames (no undefined).
- Shortens classname generation to `.#{propName}_#{propValue}__#{screenSize}` for more readability in markup.
- Moves typings to their own structure.

### CODE REVIEW

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.20-alpha.ba5981.0<br/>@codecademy/gamut-styles@2.2.12-alpha.ba5981.0<br/>@codecademy/gamut@3.3.1-alpha.ba5981.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e90e8c0973e3cc2b4aa0258--codecademy-styleguide.netlify.com>https://5e90e8c0973e3cc2b4aa0258--codecademy-styleguide.netlify.com</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e90e8c0973e3cc2b4aa0258>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.22-alpha.a26179.0<br/>@codecademy/gamut-styles@2.2.12-alpha.a26179.0<br/>@codecademy/gamut@3.4.2-alpha.a26179.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e94c692bff9afe48db79836--codecademy-styleguide.netlify.com>https://5e94c692bff9afe48db79836--codecademy-styleguide.netlify.com</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e94c692bff9afe48db79836>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.23-alpha.35604f.0<br/>@codecademy/gamut-styles@2.2.12-alpha.35604f.0<br/>@codecademy/gamut@3.4.3-alpha.35604f.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e94e63773387f399de69409--codecademy-styleguide.netlify.com>https://5e94e63773387f399de69409--codecademy-styleguide.netlify.com</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e94e63773387f399de69409>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.23-alpha.2ffcf4.0<br/>@codecademy/gamut-styles@2.2.12-alpha.2ffcf4.0<br/>@codecademy/gamut@3.4.3-alpha.2ffcf4.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e96233b9167480189335da6--codecademy-styleguide.netlify.com>https://5e96233b9167480189335da6--codecademy-styleguide.netlify.com</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e96233b9167480189335da6>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.25-alpha.f4ad46.0<br/>@codecademy/gamut@3.5.2-alpha.f4ad46.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e986869f5d2020179d0119f--codecademy-styleguide.netlify.app>https://5e986869f5d2020179d0119f--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e986869f5d2020179d0119f>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.26-alpha.e4ee44.0<br/>@codecademy/gamut@3.6.1-alpha.e4ee44.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e986cd6f31f3a017add7b42--codecademy-styleguide.netlify.app>https://5e986cd6f31f3a017add7b42--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e986cd6f31f3a017add7b42>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.26-alpha.ffc822.0<br/>@codecademy/gamut@3.6.1-alpha.ffc822.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e988251e3fe6c7b4c126ff8--codecademy-styleguide.netlify.app>https://5e988251e3fe6c7b4c126ff8--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e988251e3fe6c7b4c126ff8>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.26-alpha.e59d18.0<br/>@codecademy/gamut@3.6.1-alpha.e59d18.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e988a26cdee01fecdd338fb--codecademy-styleguide.netlify.app>https://5e988a26cdee01fecdd338fb--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e988a26cdee01fecdd338fb>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.26-alpha.1c94cb.0<br/>@codecademy/gamut@3.6.1-alpha.1c94cb.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e988ee233fb13482e396416--codecademy-styleguide.netlify.app>https://5e988ee233fb13482e396416--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e988ee233fb13482e396416>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.26-alpha.69dd8c.0<br/>@codecademy/gamut@3.6.1-alpha.69dd8c.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e98918eabc14e696c842d9c--codecademy-styleguide.netlify.app>https://5e98918eabc14e696c842d9c--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e98918eabc14e696c842d9c>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.32-alpha.25ac97.0<br/>@codecademy/gamut@5.0.1-alpha.25ac97.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e99c231d3d7c622027409b4--codecademy-styleguide.netlify.app>https://5e99c231d3d7c622027409b4--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e99c231d3d7c622027409b4>Deploy Logs</a></p>

File: packages/gamut/src/utils/generateClassnames.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * Takes a prop/style configuration and templates the classes
+ * to match the styleMaps classes returning a concatenated classname string
+ */
+
+import {
+  ResponsiveProperty,
+  OptionalResponsiveProperty,
+} from '../typings/responsive-properties';
+
+export const generateClassnames = (
```
Comment (JoshuaKGoldberg) [inline packages/gamut/src/utils/generateClassnames.ts:11]: Now that this is a more general utility, can the name be something like `generateResponsiveClassnames`?

File: packages/styleguide/stories/Atoms/Typography.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+      <Heading as="h2" fontSize="lg">
+        Text
+      </Heading>
+      <p>
+        This is a single primitive to handle all <strong>{'<p>'}</strong> and{' '}
+        <strong>{'<span>'}</strong> tags with defined size intervals
+      </p>
+      {textSizes.map(size => {
+        return (
+          <>
```
Comment (JoshuaKGoldberg) [inline packages/styleguide/stories/Atoms/Typography.stories.tsx:73]: keeeeyyyyy

File: packages/styleguide/stories/Atoms/ProgressBar/ProgressBar.stories.tsx [hunk 3]
```diff
@@ -46,11 +46,7 @@ export const progressBar = () => (
       {[0, 25, 50, 75, 100].map(percent =>
         bars.map(({ displayLabel, style }) => (
           <Column key={[percent, displayLabel, style].join()} size={6}>
-            <ProgressBar
-              displayLabel={displayLabel}
-              percent={percent}
-              style={style}
-            />
+            <ProgressBar percent={percent} style={style} />
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/styleguide/stories/Atoms/ProgressBar/ProgressBar.stories.tsx:49]: https://github.com/Codecademy/client-modules/pull/735

File: packages/gamut/src/utils/generateClassnames.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  config: ResponsiveProperty<unknown> | OptionalResponsiveProperty<unknown>,
+  styleMap: Record<string, string>
+): string[] => {
+  const styleList: string[] = [];
```
Comment (JoshuaKGoldberg) [inline packages/gamut/src/utils/generateClassnames.ts:15]: If you wanted, you could turn this whole thing into a `.map`/`.filter(Boolean)` ✨

File: packages/gamut/src/utils/generateClassnames.ts (hunk 1 above)
Comment (codecaaron) [reply packages/gamut/src/utils/generateClassnames.ts:11]: Good call, more declarative :D__

File: packages/styleguide/stories/Atoms/Typography.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+const textSizes: ('sm' | 'lg')[] = ['lg', 'sm'];
+
+export const baseTypography = () => (
+  <LayoutGrid rowGap="lg">
+    <Column size={12}>
+      <StoryTemplate status={StoryStatus.InProgress}>
+        <StoryDescription>
+          Responsive Typography primitives for all your textual needs.
+        </StoryDescription>
+      </StoryTemplate>
```
[the remark this answers is not in the document]
Comment (codecaaron) [reply packages/styleguide/stories/Atoms/Typography.stories.tsx]: 👍

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.32-alpha.3cfa3f.0<br/>@codecademy/gamut@5.0.1-alpha.3cfa3f.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e9a08e26446ed01fe98856c--codecademy-styleguide.netlify.app>https://5e9a08e26446ed01fe98856c--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e9a08e26446ed01fe98856c>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@0.8.33-alpha.4bf6d9.0<br/>@codecademy/gamut@5.1.1-alpha.4bf6d9.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e9dd092992f028b02944664--codecademy-styleguide.netlify.app>https://5e9dd092992f028b02944664--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e9dd092992f028b02944664>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@1.0.1-alpha.8fe519.0<br/>@codecademy/gamut@6.1.0-alpha.8fe519.0<br/></pre>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@1.0.2-alpha.a17c32.0<br/>@codecademy/gamut@6.1.1-alpha.a17c32.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e9de4575d3dcf508f4c5da3--codecademy-styleguide.netlify.app>https://5e9de4575d3dcf508f4c5da3--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e9de4575d3dcf508f4c5da3>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@1.0.2-alpha.4fa80b.0<br/>@codecademy/gamut@6.1.1-alpha.4fa80b.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e9ee32c0fe1c27ebccf7acc--codecademy-styleguide.netlify.app>https://5e9ee32c0fe1c27ebccf7acc--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e9ee32c0fe1c27ebccf7acc>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e9ee47a570b1b44fc3b207d--codecademy-styleguide.netlify.app>https://5e9ee47a570b1b44fc3b207d--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e9ee47a570b1b44fc3b207d>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@1.0.2-alpha.209d53.0<br/>@codecademy/gamut@6.1.1-alpha.209d53.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e9ee8b059b4328a43242b40--codecademy-styleguide.netlify.app>https://5e9ee8b059b4328a43242b40--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e9ee8b059b4328a43242b40>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@1.0.2-alpha.cac7a1.0<br/>@codecademy/gamut@6.1.1-alpha.cac7a1.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5e9eefa759b432fd00242b47--codecademy-styleguide.netlify.app>https://5e9eefa759b432fd00242b47--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5e9eefa759b432fd00242b47>Deploy Logs</a></p>

**Pull request merged**